ENH: Added iqr function to compute IQR metric in scipy/stats/stats.py

madphysicist opened the pull request:
Transfer of numpy PR https://github.com/numpy/numpy/pull/7137.

This function is different from the IQR in statsmodels. It appears to have different inputs (one array versus two), which actually makes this version more general.

As per @rgommers's [request][1], I have added `nan_policy`, which basically just selects between `np.percentile` and `np.nanpercentile`. Since there were some bugs in the shape of arrays returned by `np.nanpercentile`, I have added a workaround that should eventually be removed.

As per @josef-pkt's [request][2], I have added a 'scale' argument that accepts the values `'raw'` or `1.0`, `'normal'` or `sp.special.erfinv(0.5) * 2 * sp.sqrt(2) ~= 1.349`, or any other number. The result is divided by the scale. The default is `'raw'`.

  [1]: https://github.com/numpy/numpy/pull/7137#issuecomment-176633198
  [2]: https://github.com/numpy/numpy/pull/7137#issuecomment-176522779

The tests have been transferred almost as-is from numpy. However, if the numpy version being used is < 1.11.0, tests involving `interpolation='midpoint'` are turned off because bug-fix https://github.com/numpy/numpy/pull/7129 and its backport https://github.com/numpy/numpy/pull/7166 would not be in effect.

rgommers commented on doc/release/0.18.0-notes.rst:
```diff
@@ -37,6 +37,12 @@ This class represents a piecewise cubic polynomial passing through given points
+`scipy.stats.iqr` function
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Computes the interquatrile region of a distribution.
```
typo "interquartile"

rgommers commented on scipy/stats/stats.py:
```diff
@@ -82,6 +82,9 @@
     sem
+    zmap
+    zscore
+    iqr
```
this addition is fine, but the more important one should be made in the function listing in ``stats/__init__.py``

rgommers replied:
this is done now. deleting the comment to make it not show up in the PR conversation

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+    The concept for this function is borrowed heavily from the `numpy.histogram`
+    helper function `_hist_optim_numbins_estimator`.
+    """
+    assert isinstance(scale, string_types)
```
No ``assert`` please, because this line may disappear after byte-compiling with ``-O``. You want

    if not isinstance(scale, string_types):
        raise informative_error

madphysicist replied:
This is taken fromhttps://github.com/numpy/numpy/blame/master/numpy/lib/function_base.py#L88. Should I make the same change in numpy then?

madphysicist replied:
I am also thinking that the assert should just be removed in both cases, given how the function is called.

rgommers replied:
Yes please, and yes. I commented on the removal in the calling function already.

madphysicist replied:
See https://github.com/numpy/numpy/pull/7193.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+        raise ValueError("{0} not a valid scale for `iqr`".format(scale))
+
+
+def iqr(x, axis=None, range=(25, 75), scale='raw', nan_policy='propagate',
```
``range`` is a builtin, so can you rename this keyword?

madphysicist replied:
Do you prefer `rng` or `interval`? I prefer the former because it's shorter and preserves the meaning of the 'R' in IQR. I only suggest the latter because it is a real word rather than an abbreviation, which seems more consistent with the other keywords.

rgommers replied:
Either one is fine with me.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+
+    The ``range`` parameter allows this function to compute other percentile
+    ranges than the actual IQR. For example, setting ``range=(0, 100)`` is
+    equivalent to ``numpy.ptp``.
```
if you use single backticks for functions, they get turned into links. So that's preferred for function names.

madphysicist replied:
I did not know that worked for functions in numpy as well.

rgommers replied:
It uses the "inter-sphinx" functionality, which allows linking between different projects and the stdlib

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+    ranges than the actual IQR. For example, setting ``range=(0, 100)`` is
+    equivalent to ``numpy.ptp``.
+
+    The IQR of an empty array is scalar zero.
```
Wouldn't it be more with ``np.sum`` & co to return ``nan`` here?

madphysicist replied:
It is more thematically appropriate to stick with whatever `std` or `var` do. Another alternative is to raise something like percentile already does. I would like to see a more convincing argument either way before I change this.

rgommers replied:
Those do the same right?

    In [15]: np.var([])
    Out[15]: nan

    In [16]: np.std([])
    Out[16]: nan

    In [17]: np.__version__
    Out[17]: '1.10.4'

Raising is never a good idea I think, if more functions start doing that then empty arrays become unusable.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+    axis : int or sequence of int, optional
+        Axis along which the range is computed. The default is to compute the
+        IQR for the entire array.
+    range : Two-element sequence containing floats in range of [0,100]
```
``, optional``

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+        Percentiles over which to compute the range. Each must be between 0 and
+        100, inclusive. The default is the true IQR: (25, 75). The order of the
+        elements is not important.
+    scale : scalar or str
```
``, optional``

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+        values. Default is 'propagate'.
+    interpolation : {'linear', 'lower', 'higher', 'midpoint', 'nearest'}
+        This optional parameter specifies the interpolation method to use
+        when the percentile boundaries lie between two data points `i` and `j`:
```
didn't build the docs, but I think this needs a blank line for the list to render correctly

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+    References
+    ----------
+    .. [1] "Interquatrile range" https://en.wikipedia.org/wiki/Interquartile_range
+    .. [2] "Interquatrile range" http://www.stattrek.com/statistics/dictionary.aspx?definition=Interquartile%20range
```
same typo in [1] and [2] (interquatrile)

rgommers replied:
do we need the second link? It is a not too popular website, chances are that the link is broken a few years from now

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+
+    See Also
+    --------
+    std, var
```
``numpy.std``, ``numpy.var``

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+    """
+    x = asarray(x)
+
+    # Possibly throw an exception here, so fail-fast, before lengthy computation
```
this comment sound like a todo, and the check below is double with the ``assert`` I commented on in ``_iqr_scale_to_num``. It's fine to do the input validation inside ``_iqr_scale_to_num`` and remove it here.

madphysicist replied:
I think the intention of the comment is fine, but it uses terrible wording. "Possibly throw an exception here" should read "An exception may be thrown here".

madphysicist replied:
I changed the comment but am willing to remove it entirely if you think it is superfluous. It is a note-to-self that may or may not be useful for future maintenance.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+
+    # This check prevents percentile from raising an error later. Also, it is
+    # the right thing to do.
+    if not x.size: return 0.0
```
PEP8 nitpick: line break after ``x.size:``

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+
+    # reverse is True because the original computation was a one-liner:
+    #    return np.subtract(*percentile_func(x, sorted(range, reverse=True), ...), out=out)
+    # The current version is a little less obfscated
```
this comment seems to serve no purpose anymore

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+                          interpolation=interpolation, keepdims=keepdims)
+    out = np.subtract(pct[0], pct[1])
+
+    if scale != 1.0:
```
this ``if`` clause isn't needed, just divide in all cases

madphysicist replied:
Wouldn't that cause an unnecessary slowdown for the default case for large data sets?

rgommers replied:
Hmm, you're right. And comparing to float is not too bad here - will usually work and not break if it does't. So ignore my comment.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,186 @@ def zmap(scores, compare, axis=0, ddof=0):
+    result = np.nanpercentile(x, q, axis=axis,
+                              interpolation=interpolation, keepdims=keepdims)
+    # If non-scalar result and nanpercentile does not do proper axis roll
+    if result.ndim > 1 and NumpyVersion(np.__version__) > '1.11.0b3':
```
I think you mean ``< '1.11.0b3':``?

madphysicist commented:
I have addressed all of your comments. Additionally, I wrapped all the docs to 72 characters except a couple of lines where that did not seem appropriate, and removed the `reversed` keyword to `sort` along with the pointless comment above it.

madphysicist commented:
I am trying to think of how to handle older versions of numpy. Would a warning that "interpolation will be ignored as it is not supported for numpy versions < 1.09" be acceptable/sufficient?

madphysicist commented:
I am also not sure what is wrong with the formatting of the list I have in the docstring, but it seems to be causing a warning that is turning into an error? Could you point me to a reference on the subject please?

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2306,6 +2311,204 @@ def zmap(scores, compare, axis=0, ddof=0):
+    if NumpyVersion(np.__version__) < '1.9.0':
+        if interpolation != 'linear':
+            warnings.warn("interpolation and keepdims keywords not supported "
+                          "for numpy versions < 1.0.9")
```
typo, should be ``1.9.0``

rgommers commented:
There are a number of test failures, most of which don't look related to numpy version.

> I am trying to think of how to handle older versions of numpy. Would a warning that "interpolation will be ignored as it is not supported for numpy versions < 1.09" be acceptable/sufficient?

That seems acceptable to me.

madphysicist commented:
I am pretty sure that I have traced down (but not fixed) all the remaining failures, and they are in fact all tied to the numpy version:

  - For the numpy 1.7.x case, `np.nanpercentile` simply does not exist, so I will have to fail over to `scipy.stats.scoreatpercentile`.
  - For the numpy 1.9.x case, I did a diff of `np.nanpercentile` between `upstream/maintenance/1.9.x` and `upstream/maintenance/1.11.x` branches. Turns out the indexing for multiple dimensions changed. I will have to handle this case separately as well.

The first few tests pass because they rely on the latest versions of numpy. Basically, this function will do some strange things for `numpy` versions from `1.9.0` to `1.11.0` and start dropping features below that.

rgommers commented:
thanks, makes sense

madphysicist commented:
I finally found where NaN handling for `percentile` was corrected. The original issue was https://github.com/numpy/numpy/issues/586 and the fix was https://github.com/numpy/numpy/pull/5753. This means that `NaN` handling is neither exactly 'omit' nor 'propagate' for `np.__version__ < 1.11.0b3`. NaNs get moved to the end and can either skew the result or set it to NaN. A warning will have to suffice here as well. Here is an example of the issue for `np.__version__ == 1.9.1`:

    >>> x = np.arange(15.0)
    >>> x[1, 2] = np.nan
    >>> x
    array([  0.,   1.,   2.,   3.,   4.,   5.,   6.,  nan,   8.,   9.,  10.,
            11.,  12.,  13.,  14.])
    >>> np.percentile(x, 25)
    3.5 # Expecting 3.25, but NaN shifted result
    >>> np.percentile(x, 99)
    nan # Expected 13.867

This comment is provided so that someone looking at the soon-to-be-completed code understands the insanity it commits with the `NumpyVersion` object.

madphysicist commented:
For reference about some of the travesties I am committing, see the aforementioned numpy PR, especially starting with comment https://github.com/numpy/numpy/pull/5753#issuecomment-184854966.

madphysicist commented on scipy/stats/stats.py:
```diff
@@ -2315,6 +2320,275 @@ def zmap(scores, compare, axis=0, ddof=0):
+            else:
+                # roll multiple axes to the end and flatten that part out
+                for ax in axis[::-1]:
+                    x = np.rollaxis(x, ax, x.ndim)
```
To support multiple axes (without `keepdims`), I independently implemented a subset of `np.function_base._ureduce`. The main difference is that `_ureduce` uses `x.swapaxes` while I use `np.rollaxis`. The functionality is effectively the same since all the reduction axes get flattened out. Besides where the axes end up, is there any difference between the two methods (such as performance) that I should be aware of?

madphysicist commented:
One last version mismatch remaining... So close yet so far away.

madphysicist commented:
@rgommers I believe that this PR is complete and ready for re-review. The basic functionality remains unchanged, but I added a lot of code to handle changes to numpy.

madphysicist commented:
I am wondering how much of scipy has to jump through the same hoops to build on top of numpy successfully.

argriffing commented:
> I am wondering how much of scipy has to jump through the same hoops

Here's an example where I gave up, and this was back when there were fewer hoops.
https://github.com/scipy/scipy/pull/3610

madphysicist commented:
@argriffing I checked out your PR. It appears much more complex than the case I am working with. This is the problem with building on an arbitrary version of numpy that may not have the required features. It looks like you had to reimplement quite a bit of the functionality. In my case, I think that I was able to do it successfully.

madphysicist commented:
Squawk

argriffing commented:
This PR is aesthetically scary but I hope it can be merged.  I'm not quite brave enough to merge this one unilaterally.

Eric89GXL commented on scipy/stats/stats.py:
```diff
@@ -2312,6 +2318,275 @@ def zmap(scores, compare, axis=0, ddof=0):
+    try:
+        result = np.percentile(x, q, axis=axis, keepdims=keepdims,
+                               interpolation=interpolation)
+    except TypeError:
```
Can you comment on the circumstances under which this will happen? It will help the next dev understand what can give rise to this and what needs to be modified.

Eric89GXL commented on scipy/stats/stats.py:
```diff
@@ -2312,6 +2318,275 @@ def zmap(scores, compare, axis=0, ddof=0):
+                x = x.reshape(x.shape[:-axis.size] +
+                              (np.prod(x.shape[-axis.size:]),))
+                axis = -1
+        except TypeError:
```
Does all of this code need to be in `try/except`? Again can you mention what you expect to give rise to the `TypeError`?

Eric89GXL commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1424,6 +1425,372 @@ def test_zscore_ddof(self):
         assert_array_almost_equal(z[1], z1_expected)
 
 
+class _numpy_version_warn_context_mgr:
```
(object)

Eric89GXL commented on scipy/stats/stats.py:
```diff
@@ -2312,6 +2318,275 @@ def zmap(scores, compare, axis=0, ddof=0):
+
+
+def iqr(x, axis=None, rng=(25, 75), scale='raw', nan_policy='propagate',
+           interpolation='linear', keepdims=False):
```
PEP8 says this should align to the opening `(`

Eric89GXL commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1424,6 +1425,372 @@ def test_zscore_ddof(self):
+    contains expected number of warnings. 
+    """
+    assert_equal(len(warn_list), expected_len, "number of warnings")
+    for i in range(len(warn_list)):
```
Why not iterate directly over `for warn_ in warn_list`?

Eric89GXL commented on scipy/stats/stats.py:
```diff
@@ -2312,6 +2318,275 @@ def zmap(scores, compare, axis=0, ddof=0):
+    try:
+        return scale_conversions[scale.lower()]
+    except KeyError:
+        raise ValueError("{0} not a valid scale for `iqr`".format(scale))
```
I don't like the use of `try/except` instead of `if key not in x` checks is because on Python3 this will have an extra-long traceback, where it says an error was thrown during the handling of another error.

It also seems a little overkill to have this in a separate function given how simple it is (and it's only used in one place).

Eric89GXL commented:
Other than my nitpicks LGTM. @argriffing I see what you mean about the aesthetics, the `numpy` bug/version stuff makes it tough. But it is well documented at least, so I can't think of a cleaner solution.

argriffing commented:
Thanks for looking at this, @Eric89GXL. I'm OK with this being merged once your comments have been addressed.

argriffing commented:
@madphysicist ping

madphysicist commented:
Sorry for the delay. I'm just getting back to work after having kid \#2. I will get to addressing the issues pretty soon. The PR is not dead and I have not forgotten about it.

rgommers commented:
Congrats @madphysicist!

madphysicist commented:
I have addressed all the comments via appropriate code changes. While I personally have a preference for using `try...except KeyError` for dictionary lookups despite the lengthy error tracebacks, I have bowed to public opinion and used `if...not in...` instead (this has been a point of contention in a couple of my numpy PRs as well).

The aesthetics of this function are a good indication of the fact that it belongs in numpy rather than here. Perhaps if I ever get around to implementing my weighted percentile function, that will be enough motivation to allow the move.

ev-br commented:
Needs a rebase

ev-br commented:
All right, it's been okayed by Alex and Eric, so I rebased and merged it in https://github.com/scipy/scipy/commit/8346eea7e86fd4a379f473c1900d5928e29e3288
Thanks @madphysicist, all

madphysicist commented:
Thanks @ev-br. Sorry I did not respond earlier. Been offline for a couple of months since kid #2 was born.

ev-br commented:
Congratulations @madphysicist !
